File: src/main/java/io/austinzhu/type/functor/Monad.java
package io.austinzhu.type.functor;

import io.austinzhu.type.Function;
import io.austinzhu.type.Type;

public interface Monad<M extends Type<M> & Functor<M, A> & Applicative<M, A> & Monad<M, A>, A extends Type<A>> {

    Monad<M, A> ret(A a);

    <B extends Type<B>, MB extends Monad<MB, B> & Type<MB> & Functor<MB, B> & Applicative<MB, B>, MF extends Monad<MF, Function<A, MB>> & Type<MF> & Functor<MF, Function<A, MB>> & Applicative<MF, Function<A, MB>>>
    Monad<MB, B> compose(Monad<M, A> m, Monad<MF, Function<A, MB>> f);
}

File: src/main/scala/dev/austinzhu/pattern/functor/Monad.scala
package dev.austinzhu.pattern.functor

type.Function

type.Type

trait Monad[M <: Type with Functor[M, A] with Applicative[M, A] with Monad[M, A], A <: Type] {
  def ret(a: A)

  def compose[B <: Type, MB <: Monad[MB, B] with Type with Functor[MB, B] with Applicative[MB, B], MF <: Monad[MF, Function[A, MB]] with Type with Functor[MF, Function[A, MB]] with Applicative[MF, Function[A, MB]]](m: Monad[M, A], f: Monad[MF, Function[A, MB]])
}